### PR TITLE
Fix/registry basic auth

### PR DESCRIPTION
**Pull request by ycliuhw**
*This PR changes to use same base URL for dockerhub registry API v1 and v2;*

Note: It would be better to review https://github.com/juju/juju/pull/13281 first.

## Checklist

 - [ ] ~Requires a [pylibjuju](https://github.com/juju/python-libjuju) change~
 - [ ] ~Added [integration tests](https://github.com/juju/juju/tree/develop/tests) for the PR~
 - [ ] ~Added or updated [doc.go](https://discourse.jujucharms.com/t/readme-in-packages/451) related to packages changed~
 - [x] Comments answer the question of why design decisions were made

## QA steps

```console
# copy your dockerhub credentials from ~/.docker/config.json
$ cat tttt.json
{
    // "auth": "xxxxxx==",
    "username": "a",
    "password": "pwd",
    "repository": "ycliuhw"
}


$ JUJU_DEV_FEATURE_FLAGS=private-registry juju bootstrap microk8s k1 --config caas-image-repo="'$(cat tttt.json)'"

$ mkubectl -ncontroller-k1 get pod/controller-0 -o json | jq .spec.imagePullSecrets
[
  {
    "name": "juju-image-pull-secret"
  }
]

$ mkubectl -ncontroller-k1 get secret/juju-image-pull-secret -o json | jq -r '.data[".dockerconfigjson"]' | base64 --decode | jq
{
  "auths": {
    "https://index.docker.io/v1/": {
      "username": "a",
      "password": "pwd",
      "auth": "xxxxxx=="
    }
  }
}

```

## Documentation changes

Yes

## Bug reference

https://bugs.launchpad.net/juju/+bug/1935830

### CODE REVIEW

File: docker/auth.go [hunk 1]
```diff
@@ -70,7 +75,13 @@ func (ba BasicAuthConfig) Empty() bool {
...
 
 // Validate validates the spec.
-func (ba BasicAuthConfig) Validate() error {
+func (ba *BasicAuthConfig) Validate() error {
+	if ba.Empty() {
+		return nil
+	}
+	if ba.Auth == "" {
+		ba.Auth = base64.StdEncoding.EncodeToString([]byte(ba.Username + ":" + ba.Password))
+	}
```
Comment (wallyworld) [inline docker/auth.go:84]: Validate() shouldn't have a side effect. Can we set Auth elsewhere, eg when the struct is created or something

File: docker/registry/registry_test.go [hunk 2]
```diff
@@ -65,7 +65,7 @@ func (s *registrySuite) getRegistry(c *gc.C) (registry.Registry, *gomock.Control
 				func(req *http.Request) (*http.Response, error) {
 					c.Assert(req.Header, jc.DeepEquals, http.Header{"Authorization": []string{"Bearer xxxxx=="}})
 					c.Assert(req.Method, jc.DeepEquals, `GET`)
-					c.Assert(req.URL.String(), jc.DeepEquals, `https://quay.io`)
+					c.Assert(req.URL.String(), jc.DeepEquals, `https://quay.io/v2`)
```
Comment (wallyworld) [inline docker/registry/registry_test.go:68]: `DeepEquals` for a string seems extreme :-)

**Review by wallyworld — APPROVED** [review]

**Comment by ycliuhw** [comment]
$$merge$$

**Comment by ycliuhw** [comment]
[integrate more registry support, for example gcr, azure, etc.](https://trello.com/c/K1isSDRz/1749-integrate-more-registry-support-for-example-gcr-azure-etc)

**Pull request merged**